// @agc93/privatebin/dist/encryption.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.encryptMessageBuffer = exports.EncryptionBuilder = void 0;
const crypto_1 = __importDefault(require("crypto"));
const util_1 = require("util");
const pbkdf2 = util_1.promisify(crypto_1.default.pbkdf2);
/**
 * A fluent builder to help configure encryption options.
 */
class EncryptionBuilder {
    /**
     *
     */
    constructor() {
        /* usePassword = (password: string): EncryptionBuilder => {
            this._options.password = () => password;
            return this;
        } */
        this.getDefaults = () => {
            return {
                algorithm: 'aes',
                encryptionMode: 'gcm',
                keySize: 256,
                tagSize: 128,
                iterations: 100000,
                compression: 'none',
                key: crypto_1.default.randomBytes(32)
            };
        };
        this.buildOptions = () => {
            var defaults = this.getDefaults();
            return { ...defaults, ...this._options };
        };
        this._options = {};
    }
    setKeySize(keySize) {
        this._options.keySize = keySize;
        return this;
    }
    setTagSize(tagSize) {
        this._options.tagSize = tagSize;
        return this;
    }
    setIterations(iterations) {
        this._options.iterations = iterations;
        return this;
    }
    enableCompression(enable = true) {
        this._options.compression = enable ? 'zlib' : 'none';
        return this;
    }
    useKey(key) {
        if (typeof (key) === 'string') {
            this._options.key = Buffer.from(key, 'utf8');
        }
        else {
            this._options.key = key;
        }
        return this;
    }
}
exports.EncryptionBuilder = EncryptionBuilder;
async function encryptMessageBuffer(messageBytes, encryptionOptions, uploadOptions) {
    let iv = crypto_1.default.randomBytes(16);
    let salt = crypto_1.default.randomBytes(8);
    // this should use a key on IEncryptionOptions
    // let password = crypto.randomBytes(32);
    var cryptedPassword = await generateKey(encryptionOptions.key, salt, encryptionOptions);
    var crypt = { iv, salt, key: cryptedPassword };
    var algorithm = `aes-${encryptionOptions.keySize}-${encryptionOptions.encryptionMode}`; // even though it might not actually be GCM
    var cipher = crypto_1.default.createCipheriv(algorithm, cryptedPassword, iv, { authTagLength: 16 });
    // var cipher = crypto.createCipheriv(algorithm, cryptedPassword, iv);
    var data = getData(crypt, encryptionOptions, uploadOptions);
    cipher.setAAD(Buffer.from(JSON.stringify(data), 'utf8'));
    var cipherText = Buffer.concat([cipher.update(messageBytes), cipher.final(), cipher.getAuthTag()]);
    return { cipherText, data };
}
exports.encryptMessageBuffer = encryptMessageBuffer;
async function generateKey(password, salt, options) {
    var cryptResult = await pbkdf2(password !== null && password !== void 0 ? password : crypto_1.default.randomBytes(32), salt, options.iterations, password.length, 'sha256');
    // var cryptResult = crypto.pbkdf2Sync(password, salt, options.iterations, options.keySize / 8, 'sha256');
    return cryptResult;
}
function getData(params, opts, request) {
    const adata = [];
    adata.push([params.iv.toString('base64'), params.salt.toString('base64'), opts.iterations, opts.keySize, opts.tagSize, 'aes', opts.encryptionMode, opts.compression]);
    adata.push(request.uploadFormat);
    adata.push(request.openDiscussion ? 1 : 0);
    adata.push(request.burnAfterReading ? 1 : 0);
    return adata;
}
