// @agc93/privatebin/dist/uploadClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateBinClient = void 0;
const encryption_1 = require("./encryption");
const pako_1 = __importDefault(require("pako"));
const axios_1 = __importDefault(require("axios"));
const bs58_1 = require("bs58");
/**
 * A simple client class to encapsulate uploading pastes to a PrivateBin server
 */
class PrivateBinClient {
    /**
     *  Creates a new PrivateBin client for the given server URL
     *
     * @param baseUrl: The base URL of the server. This should be the root address like `https://privatebin.net`
     */
    constructor(baseUrl) {
        this.getMessageBytes = (data, compression) => {
            const buffer = Buffer.from(JSON.stringify({ paste: data }), 'utf8');
            if (compression === 'zlib') {
                return Buffer.from(pako_1.default.deflateRaw(new Uint8Array(buffer)));
            }
            else {
                return buffer;
            }
        };
        /**
         * Uploads given content as a paste to the PrivateBin server.
         *
         * @param content The paste contents to upload
         * @param options The upload options to use
         * @param encryptOpts Additional encryption options for your paste
         */
        this.uploadContent = async (content, options, encryptOpts) => {
            var opts = { ...this.getDefaultOptions(), ...options };
            var encryptionOpts = encryptOpts !== null && encryptOpts !== void 0 ? encryptOpts : new encryption_1.EncryptionBuilder().enableCompression(true).buildOptions();
            var messageBuffer = this.getMessageBytes(content, encryptionOpts.compression);
            var encryptedData = await encryption_1.encryptMessageBuffer(messageBuffer, encryptionOpts, opts);
            var resp = await this.postPasteContent(encryptedData.data, encryptedData.cipherText, opts);
            if (resp.status == 201 || resp.status == 200) {
                return {
                    response: resp.data,
                    success: resp.data.status == 0,
                    urlKey: bs58_1.encode(encryptionOpts.key),
                    url: `${this._baseUrl}${resp.data.url}`
                };
            }
            throw new Error("Upload failed!");
        };
        this.postPasteContent = async (requestData, cipherText, options) => {
            var requestBody = {
                v: 2,
                ct: cipherText.toString('base64'),
                adata: requestData,
                meta: {
                    expire: options.expiry
                }
            };
            var resp = await axios_1.default.post("/", requestBody, this.getRequestConfig());
            return resp;
        };
        this.getRequestConfig = () => {
            return {
                baseURL: this._baseUrl,
                headers: {
                    'Content-Type': 'application/json',
                    'X-Requested-With': 'JSONHttpRequest' // required by privatebin. yikes that casing.
                }
            };
        };
        baseUrl = baseUrl.startsWith('https://') ? baseUrl : `https://${baseUrl}`;
        this._baseUrl = baseUrl.replace(/\/+$/, ""); // remove any trailing slashes
    }
    getDefaultOptions() {
        return {
            expiry: '1week',
            openDiscussion: false,
            burnAfterReading: false,
            uploadFormat: 'markdown'
        };
    }
}
exports.PrivateBinClient = PrivateBinClient;
